From 813e68f7718e233f0dae8f1a86ad5688c53dcaa7 Mon Sep 17 00:00:00 2001
From: Shuguang Gong <shuguang.gong@intel.com>
Date: Wed, 18 Mar 2015 12:45:45 +0800
Subject: [PATCH 319/323] ov2685: add fluorescent awb setting

Change-Id: I1b8fab620b6509b3e6f2e97a7ba3af14efac40dd
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-26355
Signed-off-by: Shuguang Gong <shuguang.gong@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/46780
Validation-Android: Wei Tao <wei.tao@intel.com>
Approver: Jing Tao <jing.tao@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Integrator: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../camera/drivers/media/i2c/ov2685.c                 | 19 ++++++++++++++-----
 .../camera/drivers/media/i2c/ov2685.h                 | 14 ++++++++++++++
 2 files changed, 28 insertions(+), 5 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
index 074d4ec..40582fc 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
@@ -325,27 +325,36 @@ static int ov2685_s_wb(struct v4l2_subdev *sd, int value)
 {
 	struct i2c_client *client = v4l2_get_subdevdata(sd);
 	struct ov2685_device *dev = to_ov2685_sensor(sd);
+	int ret;
 
 	switch (value) {
 	case V4L2_WHITE_BALANCE_MANUAL:
-		ov2685_write_reg_array(client, ov2685_AWB_manual);
+		ret = ov2685_write_reg_array(client, ov2685_AWB_manual);
 		break;
 	case V4L2_WHITE_BALANCE_AUTO:
-		ov2685_write_reg_array(client, ov2685_AWB_auto);
+		ret = ov2685_write_reg_array(client, ov2685_AWB_auto);
 		break;
 	case V4L2_WHITE_BALANCE_INCANDESCENT:
-		ov2685_write_reg_array(client, ov2685_AWB_incandescent);
+		ret = ov2685_write_reg_array(client, ov2685_AWB_incandescent);
+		break;
+	case V4L2_WHITE_BALANCE_FLUORESCENT:
+	case V4L2_WHITE_BALANCE_FLUORESCENT_H:
+		ret = ov2685_write_reg_array(client, ov2685_AWB_fluorescent);
 		break;
 	case V4L2_WHITE_BALANCE_CLOUDY:
-		ov2685_write_reg_array(client, ov2685_AWB_cloudy);
+		ret = ov2685_write_reg_array(client, ov2685_AWB_cloudy);
 		break;
 	case V4L2_WHITE_BALANCE_DAYLIGHT:
-		ov2685_write_reg_array(client, ov2685_AWB_sunny);
+		ret = ov2685_write_reg_array(client, ov2685_AWB_sunny);
 		break;
 	default:
 		dev_err(&client->dev, "ov2685_s_wb: %d\n", value);
+		return -EINVAL;
 	}
 
+	if (ret)
+		return ret;
+
 	dev->wb_mode = value;
 	return 0;
 }
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h
index 65f21c8..ae0afb8 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h
@@ -326,6 +326,20 @@ static struct ov2685_reg const ov2685_AWB_incandescent[] = {
 	{OV2685_TOK_TERM, 0, 0}
 };
 
+static struct ov2685_reg const ov2685_AWB_fluorescent[] = {
+	{OV2685_8BIT, 0x3208 , 0x00},
+	{OV2685_8BIT, 0x5180 , 0xf6},
+	{OV2685_8BIT, 0x5195 , 0x04},
+	{OV2685_8BIT, 0x5196 , 0x90},
+	{OV2685_8BIT, 0x5197 , 0x04},
+	{OV2685_8BIT, 0x5198 , 0x00},
+	{OV2685_8BIT, 0x5199 , 0x09},
+	{OV2685_8BIT, 0x519a , 0x20},
+	{OV2685_8BIT, 0x3208 , 0x10},
+	{OV2685_8BIT, 0x3208 , 0xa0},
+	{OV2685_TOK_TERM, 0, 0}
+};
+
 static struct ov2685_reg const ov2685_AWB_cloudy[] = {
 	{OV2685_8BIT, 0x3208 , 0x00},
 	{OV2685_8BIT, 0x5180 , 0xf6},
-- 
1.9.1

